File: app/src/main/java/com/android/cast/dlna/demo/QueryFragment.java
package com.android.cast.dlna.demo;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.android.cast.dlna.DLNACastManager;

import org.fourthline.cling.model.meta.Device;

public class QueryFragment extends Fragment implements IDisplayDevice {

    private TextView mMediaInfo;
    private TextView mPositionInfo;
    private TextView mTransportInfo;
    private TextView mVolumeInfo;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_query, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initComponent(view);
    }

    private void initComponent(View view) {
        mMediaInfo = view.findViewById(R.id.ctrl_device_query_media_info);
        mPositionInfo = view.findViewById(R.id.ctrl_device_query_position_info);
        mTransportInfo = view.findViewById(R.id.ctrl_device_query_transport_info);
        mVolumeInfo = view.findViewById(R.id.ctrl_device_query_volume_info);
        view.findViewById(R.id.ctrl_device_query_refresh).setOnClickListener((v -> setInfo()));
    }

    private Device<?, ?, ?> mDevice;

    @Override
    public void setCastDevice(Device<?, ?, ?> device) {
        mDevice = device;
        if (!isHidden()) setInfo();
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        if (!hidden) setInfo();
    }

    private void setInfo() {
        mMediaInfo.setText("");
        mPositionInfo.setText("");
        mTransportInfo.setText("");
        mVolumeInfo.setText("");

        if (mDevice != null) {
            DLNACastManager.getInstance().getMediaInfo(mDevice, (mediaInfo, errMsg) ->
                    mMediaInfo.setText(String.format("MediaInfo:\n%s", mediaInfo != null ? mediaInfo.getCurrentURI() : errMsg)));

            DLNACastManager.getInstance().getPositionInfo(mDevice, (positionInfo, errMsg) -> {
                try {
                    mPositionInfo.setText(String.format("PositionInfo:\n%s", positionInfo != null ? positionInfo : errMsg));
                } catch (Exception e) {
                    e.printStackTrace();
                    mPositionInfo.setText(e.toString());
                }
            });

            DLNACastManager.getInstance().getTransportInfo(mDevice, (transportInfo, errMsg) ->
                    mTransportInfo.setText(String.format("TransportInfo:\n%s", transportInfo != null ? transportInfo.getCurrentTransportState() : errMsg)));

            DLNACastManager.getInstance().getVolumeInfo(mDevice, (volume, errMsg) ->
                    mVolumeInfo.setText(String.format("Volume: %s", volume != null ? volume : errMsg)));
        }
    }
}

File: app/src/main/java/com/android/cast/dlna/demo/QueryFragment.kt
package com.android.cast.dlna.demo

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.android.cast.dlna.core.ContentType.VIDEO
import com.android.cast.dlna.dmc.DLNACastManager
import org.fourthline.cling.model.meta.Device
import org.fourthline.cling.support.model.DIDLContent
import org.fourthline.cling.support.model.MediaInfo
import org.fourthline.cling.support.model.PositionInfo
import org.fourthline.cling.support.model.TransportInfo
import org.fourthline.cling.support.model.item.Item

class QueryFragment : Fragment(), IDisplayDevice {

    private val mediaInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_device_query_media_info) }
    private val positionInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_device_query_position_info) }
    private val transportInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_device_query_transport_info) }
    private val volumeInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_device_query_volume_info) }
    private val browseInfo: TextView? by lazy { view?.findViewById(R.id.ctrl_device_browse_info) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_query, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initComponent(view)
    }

    private fun initComponent(view: View) {
        view.findViewById<View>(R.id.ctrl_device_query_refresh).setOnClickListener { setInfo() }
    }

    private var device: Device<*, *, *>? = null
    override fun setCastDevice(device: Device<*, *, *>?) {
        this.device = device
        if (!isHidden) setInfo()
    }

    override fun onHiddenChanged(hidden: Boolean) {
        if (!hidden) setInfo()
    }

    private fun setInfo() {
        if (device != null) {
            DLNACastManager.getInstance().getMediaInfo(device) { mediaInfo: MediaInfo?, errMsg: String? ->
                this.mediaInfo?.text = String.format("MediaInfo:\n%s", if (mediaInfo != null) mediaInfo.currentURI else errMsg)
            }
            DLNACastManager.getInstance().getPositionInfo(device) { positionInfo: PositionInfo?, errMsg: String? ->
                try {
                    this.positionInfo?.text = String.format("PositionInfo:\n%s", positionInfo ?: errMsg)
                } catch (e: Exception) {
                    e.printStackTrace()
                    this.positionInfo?.text = e.toString()
                }
            }
            DLNACastManager.getInstance().getTransportInfo(device) { transportInfo: TransportInfo?, errMsg: String? ->
                this.transportInfo?.text = String.format("TransportInfo:\n%s", if (transportInfo != null) transportInfo.currentTransportState else errMsg)
            }
            DLNACastManager.getInstance().getVolumeInfo(device) { volume: Int?, errMsg: String? ->
                volumeInfo?.text = String.format("Volume: %s", volume ?: errMsg)
            }
            DLNACastManager.getInstance().getContent(device, VIDEO) { s: DIDLContent?, errMsg: String? ->
                browseInfo?.text = if (s != null) parseContentString(s) else errMsg
            }
        } else {
            mediaInfo?.text = ""
            positionInfo?.text = ""
            transportInfo?.text = ""
            volumeInfo?.text = ""
            browseInfo?.text = ""
        }
    }

    private fun parseContentString(content: DIDLContent?): String {
        if (content == null || content.items.isEmpty() && content.containers.isEmpty()) return ""
        val builder = StringBuilder()
        if (content.containers.isNotEmpty()) {
            for (container in content.containers) {
                builder
                    .append("\nContainer: ")
                    .append(container.title)
                    .append("\nItems:\n")
                    .append(parseItems(container.items))
            }
        } else if (content.items.isNotEmpty()) {
            builder
                .append("\nItems:\n")
                .append(parseItems(content.items))
        }
        return builder.toString()
    }

    private fun parseItems(list: List<Item>?): String {
        if (list == null || list.isEmpty()) return "[]"
        val builder = StringBuilder()
        for (item in list) {
            builder.append(item.firstResource.value).append("\n")
        }
        return builder.toString()
    }
}